indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java

```diff
@@ -773,4 +806,88 @@ public void setInvalidRowCount(long invalidRowCount)
       this.invalidRowCount = invalidRowCount;
     }
   }
+
+  public static class IndexGeneratorAggregatingMapper extends IndexGeneratorMapper
+  {
+    private static final int MIN_CHECK_INTERVAL = 1024;
+
+    private final Map<Long, IncrementalIndex> combiners = Maps.newHashMap();
+
+    private int counter;
+    private int checkInterval;
+    private int maxRowsInMemory;
+
+    @Override
+    protected void setup(Context context) throws IOException, InterruptedException
+    {
+      super.setup(context);
+      maxRowsInMemory = config.getSchema().getTuningConfig().getRowFlushBoundary();
+      checkInterval = Math.max(maxRowsInMemory >> 3, MIN_CHECK_INTERVAL);
+    }
+
+    @Override
+    protected void cleanup(Context context) throws IOException, InterruptedException
+    {
+      super.cleanup(context);
+      for (IncrementalIndex index : combiners.values()) {
+        flush(context, index);
+      }
+      combiners.clear();
+    }
+
+    @Override
+    protected void innerMap(InputRow inputRow, Object value, Context context)
+        throws IOException, InterruptedException
+    {
+      final long timestamp = granularitySpec.getQueryGranularity().truncate(inputRow.getTimestampFromEpoch());
```

Comment:
How about use bucket intervals to make each IncrementalIndex to reduce memory footprint(maybe)? like the below.
`final Interval interval = config.getGranularitySpec().bucketInterval(bucket.time).get();`